README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![img](https://img.shields.io/badge/Lifecycle-Stable-97ca00)](https://github.com/bcgov/repomountie/blob/master/doc/lifecycle-badges.md)
\[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)

# tidyhydat.ws

## Project Status

This package is maintained by the Data Science and Analytics Branch of
the [British Columbia Ministry of Citizens’
Services](https://www2.gov.bc.ca/gov/content/governments/organizational-structure/ministries-organizations/ministries/citizens-services).

## Installation

``` r
#install.packages("remotes")
remotes::install_github("bcgov-c/tidyhydat.ws")
```

## Water Office web service - `realtime_ws()`

The National Hydrological Service has recently introduced an efficient
service from which to query real-time data. The `realtime_ws()`
function, which provides a convenient way to import this data into R,
introduces two new arguments that impact the data that is returned. The
web service provides additional data beyond simply hydrometric
information. This is specified using the `parameters` argument as an
integer code. The corresponding parameters can be examined using the
internal `param_id` dataset:

``` r
library(tidyhydat.ws)
data("param_id")
param_id
#>   Parameter Code                     Name_En                            Name_Fr
#> 1        46   HG     Water level provisional            Niveau d'eau temporaire
#> 2        16  HG2       Secondary water level            Niveau d'eau secondaire
#> 3        52  HG3        Tertiary water level             Niveau d'eau tertiaire
#> 4        47   QR       Discharge Provisional                D<e9>bit temporaire
#> 5         8  QRS           Discharge, sensor                  D<e9>bit, capteur
#> 6         5   TW           Water temperature            Temp<e9>rature de l'eau
#> 7        41  TW2 Secondary water temperature Temp<e9>rature de l'eau secondaire
#> 8        18   PC   Accumulated precipitation    Pr<e9>cipitations accumul<e9>es
#>   Unit
#> 1    m
#> 2    m
#> 3    m
#> 4 m3/s
#> 5 m3/s
#> 6 degC
#> 7 degC
#> 8   mm
#>                                                                                         Description_En
#> 1 Height, stage, elevation of the water surface at a specified station above some arbitrary zero datum
#> 2                     Height, stage, additional secondary water level reading at the specified station
#> 3                      Height, stage, additional tertiary water level reading at the specified station
#> 4                                                   Discharge - flow measured at points along a stream
#> 5                                                                       s Discharge measured by sensor
#> 6                                               Temperature, temperature of water at a specified depth
#> 7                 Temperature, additional secondary water temperature reading at the specified station
#> 8      Precipitation, accumulator - summation of precipitation collected beginning at a specified date
#>                                                                                                                                                Description_Fr
#> 1 Hauteur, niveau, <e9>l<e9>vation de la surface de l'eau <e0> une station sp<e9>cifi<e9>e au-dessus du z<e9>ro d'un syst<e8>me de r<e9>f<e9>rence arbitraire
#> 2                                                           Hauteur, niveau, lecture secondaire additionnelle du niveau d'eau <e0> la station sp<e9>cifi<e9>e
#> 3                                                            Hauteur, niveau, lecture tertiaire additionnelle du niveau d'eau <e0> la station sp<e9>cifi<e9>e
#> 4                                                                                 D<e9>bit - <e9>coulement mesur<e9> <e0> des points le long d'un cours d'eau
#> 5                                                                                                                              D<e9>bit mesur<e9> par capteur
#> 6                                                                                 Temp<e9>rature, temp<e9>rature de l'eau <e0> une profondeur sp<e9>cifi<e9>e
#> 7                                              Temp<e9>rature, lecture secondaire additionnelle de la temp<e9>rature de l'eau <e0> la station sp<e9>cifi<e9>e
#> 8                                                   Pr<e9>cipitations, accumulateur - somme des pr<e9>cipitations recueillies depuis une date sp<e9>cifi<e9>e
```

The `parameters` argument will take any value in the
`param_id$Parameter` column. The web service requires credentials to
access which can only be requested from ECCC. To retrieve data in this
manner, `tidyhydat.ws` employs a two stage process whereby we get a
token from the web service using our credentials then use that token to
access the web service. Therefore the second new argument is `token` the
value of which is provided by `token_ws()`:

``` r
## Get token
token_out <- token_ws()

## Input station_number, parameters and date range
ws_test <- realtime_ws(station_number = "08LG006",
                       parameters = c(46,5), ## Water level and temperature
                       start_date = "2017-06-25",
                       end_date = "2017-07-24",
                       token = token_out)
```

Tokens only last for 10 minutes and users can only have 5 tokens at
once. Therefore it is best to query the web service a little as possible
by being efficient and strategic with your queries. `realtime_ws()` will
only return data that is available. A message is returned if a
particular station was not available. `parameters`, `start_date` and
`end_date` fail silently if the station does not collect that parameter
or data on that date. The web service constrains queries to be under 60
days and fewer than 300 stations. If more data is required, multiple
queries should be made and combined using a function like `rbind()` or
`dplyr::bind_rows()`.

### Managing your credentials in R

Because you are accessing the web service using credentials and
potentially will be sharing your code will others, it is important that
you set up some method so that your secrets aren’t shared widely. Please
read [this article](http://httr.r-lib.org/articles/secrets.html) to
familiarize yourself with credential management. [This
section](http://httr.r-lib.org/articles/secrets.html#environment-variables)
is summarized here specific to `tidyhydat.ws`. If you receive your
credentials from ECCC it not advisable to directly include them in any
code. Rather these important values are again stored in the `.Renviron`
file. Run the following in a console:

``` r
file.edit("~/.Renviron")
```

This opens your `.Renviron` file which where you enter the credentials
given to you by ECCC. The code that you paste into the `.Renviron` file
should look like something like this:

``` r
## Credentials for ECCC web service
WS_USRNM = "here is the username that ECCC gave you"
WS_PWD = "here is the password that ECCC gave you"
```

Now these values can be accessed within an R session without giving away
your secrets (Using `Sys.getenv()`). For `token_ws()` they are called
automatically each time you use the function.

## Compare `realtime_ws` and `realtime_dd`

`tidyhydat.ws` provides two methods to download real-time data.
`realtime_ws()`, coupled with `token_ws()`, is an API client for a web
service hosted by ECCC. `realtime_dd()` provides a function to import
openly accessible .csv files from
[here](http://dd.weather.gc.ca/hydrometric/). `realtime_ws()` has
several difference to `realtime_dd()`. These include:

  - *Speed*: `realtime_ws()` is much faster for larger queries
    (i.e. many stations). For single station queries `realtime_dd()` if
    more appropriate.
  - *Length of record*: `realtime_ws()` records goes back further though
    only two months of data can accessed at one time. Though it varies
    for each station, typically the last 18 months of data are available
    with the web service.  
  - *Type of parameters*: `realtime_dd()` is restricted to river flow
    (either LEVEL and FLOW) data. In contrast `realtime_ws()` can
    download several different parameters depending on what is available
    for that station. See `data("param_id")` for a list and explanation
    of the parameters.
  - *Date/Time filtering*: `realtime_ws()` provides argument to select a
    date range. Selecting a data range with `realtime_dd()` is not
    possible until after all files have been downloaded.
  - *Accessibility*: `realtime_dd()` downloads data that openly
    accessible. `realtime_ws()` downloads data using a username and
    password which must be provided by ECCC.

## How to Contribute

If you would like to contribute to the package, please see our
[CONTRIBUTING](CONTRIBUTING.md) guidelines.

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

## License

Copyright 2017 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the “License”); you may
not use this file except in compliance with the License. You may obtain
a copy of the License at

<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


NEWS.md
tidyhydat.ws 0.1.5
=========================
* Make `realtime_ws` more flexible by allowing any numeric parameter input. 
* Expand `param_id`

tidyhydat.ws 0.1.4
=========================
* Added the ability to request times to `realtime_ws`.


tidyhydat.ws 0.1.3
=========================

### Minor improvement

* Added user agent to identify HTTP requests coming from tidyhydat.ws



tidyhydat.ws 0.1.2
=========================

### Minor improvement

* Removed 300 station limit
* Using rlang internally
* Added package level documentation
* Adding a check at package load if the password env variables are set

tidyhydat.ws 0.1.1 
=========================

### Bug Fixes

* Modify some column headings to accomodate changes to webservice
* Add time attribute to token object so that `realtime_ws()` can check if the token is expired. 

tidyhydat.ws 0.1.0 
=========================

### NEW FEATURES

* Split off from main tidyhydat package


tests/testthat.R
library(testthat)
library(tidyhydat.ws)

test_check("tidyhydat.ws")


tests/testthat/test_download_ws.R
context("Testing realtime functions")

test_that("realtime_ws returns the correct data header", {
  skip_on_cran()
  skip_on_travis()
  token_out <- token_ws()

  ws_test <- realtime_ws(station_number = "08MF005",
                         parameters = c(46),  ## Water level
                         start_date = Sys.Date(),
                         end_date = Sys.Date(),
                         token = token_out)

  expect_identical(colnames(ws_test),
                   c("STATION_NUMBER", "Date", "Name_En", "Value", "Unit", "Grade",
                     "Symbol", "Approval", "Parameter", "Code"))

  ## Turned #42 into a test
  expect_is(ws_test$Value, "numeric")
})


test_that("realtime_ws fails with incorrectly specified date",{
  skip_on_cran()
  skip_on_travis()
  token_out <- token_ws()

  expect_error(realtime_ws(station_number = "08MF005",
                           parameters = 46,
                           start_date = "01-01-2017 00:00:00",
                           token = token_out))
  expect_error(realtime_ws(station_number = "08MF005",
                           parameters = 46,
                           end_date = "01-01-2017 00:00:00",
                           token = token_out))
})

test_that("realtime_ws succeed specifying only date; no time",{
  skip_on_cran()
  skip_on_travis()
  token_out <- token_ws()

  sdate <- Sys.Date() - 1
  edate <- Sys.Date()

  output <- realtime_ws(station_number = "08MF005",
                           parameters = 46,
                           start_date = sdate,
                           end_date = edate,
                           token = token_out)

  expect_equal(as.Date(max(output$Date)), edate)
  expect_equal(as.Date(min(output$Date)), sdate)

})

test_that("realtime_ws succeed specifying  time",{
  skip_on_cran()
  skip_on_travis()
  token_out <- token_ws()

  stime <- as.POSIXlt(Sys.time()- 1E5, tz = "UTC")
  etime <- as.POSIXlt(Sys.time(), tz = "UTC")

  output <- realtime_ws(station_number = "08MF005",
                        parameters = 46,
                        start_date = stime,
                        end_date = etime,
                        token = token_out)

  expect_true(max(output$Date) <= etime)
  expect_true(min(output$Date) >= stime)

})


tests/testthat/helper-tidyhydat.ws.R
library("vcr")
invisible(vcr::vcr_configure())


R/stations_ws.R
#' Download realtime data from the ECCC web service [EXPERIMENTAL]
#'
#' Function to actually retrieve data from ECCC web service. Before using this function,
#' a token from \code{token_ws()} is needed. \code{realtime_ws} will let you know if the token has expired.
#' The maximum number of days that can be queried depends on other parameters being requested.
#' If one station is requested, 18 months of data can be requested. If you continually receiving
#' errors when invoking this function, reduce the number of observations (via station_number,
#' parameters or dates) being requested.
#'
#' @param station_number Water Survey of Canada station number.
#' @param parameters parameter ID. Can take multiple entries. Parameter is a numeric code. See \code{param_id} for options. Defaults to all parameters.
#' @param start_date Need to be in YYYY-MM-DD. Defaults to 30 days before current date.
#' @param end_date Need to be in YYYY-MM-DD. Defaults to current date.
#' @param token generate by \code{token_ws()}
#'
#'
#' @format A tibble with 6 variables:
#' \describe{
#'   \item{STATION_NUMBER}{Unique 7 digit Water Survey of Canada station number}
#'   \item{Date}{Observation date and time. Formatted as a POSIXct class as UTC for consistency.}
#'   \item{Name_En}{Code name in English}
#'   \item{Value}{Value of the measurement.}
#'   \item{Unit}{Value units}
#'   \item{Grade}{future use}
#'   \item{Symbol}{future use}
#'   \item{Approval}{future use}
#'   \item{Parameter}{Numeric parameter code}
#'   \item{Code}{Letter parameter code}
#' }
#'
#' @examples
#' \dontrun{
#' token_out <- token_ws()
#'
#' stations_ws(station_number = "08MF005", token = token_out)
#'
#' }
#' @family realtime functions
#' @export


stations_ws <- function(station_number, parameters = c(46, 16, 52, 47, 8, 5, 41, 18),
                        start_date = Sys.Date() - 30, end_date = Sys.Date(), token) {


  ## Check to see if the token is expired
  time_token <- attr(token, 'time')
  if(format(time_token + 10 * 60) < Sys.time()){
    stop("Your token has expired. Retrieve a new one using token_ws()")
  }



  ### Check date is in the right format
  #if (is.na(as.Date(start_date, format = "%Y-%m-%d")) | is.na(as.Date(end_date, format = "%Y-%m-%d"))) {
  #  stop("Invalid date format. Dates need to be in YYYY-MM-DD format")
  #}



  ## Build link for GET
  baseurl <- "https://wateroffice.ec.gc.ca/services/real_time_data/csv/inline?"
  station_string <- paste0("stations[]=", station_number, collapse = "&")
  parameters_string <- paste0("parameters[]=", parameters, collapse = "&")
  date_string <- paste0("start_date=", start_date, "%2000:00:00&end_date=", end_date, "%2023:59:59")
  token_string <- paste0("token=", token[1])

  ## paste them all together
  url_for_GET <- paste0(
    baseurl,
    station_string, "&",
    parameters_string, "&",
    date_string, "&",
    token_string
  )

  ## Get data
  get_ws <- httr::GET(url_for_GET)

  ## Give webservice some time
  Sys.sleep(1)

  if (httr::status_code(get_ws) == 403) {
    stop("403 Forbidden: the web service is denying your request. Try any of the following options:
         -Ensure you are not currently using all 5 tokens
         -Wait a few minutes and try again
         -Copy the token_ws call and paste it directly into the console
         -Try using realtime_ws if you only need water quantity data
         ")
  }

  ## Check the GET status
  httr::stop_for_status(get_ws)

  if (httr::headers(get_ws)$`content-type` != "text/csv; charset=utf-8") {
    stop("GET response is not a csv file")
  }

  ## Turn it into a tibble and specify correct column classes
  csv_df <- httr::content(
    get_ws,
    type = "text/csv",
    encoding = "UTF-8",
    col_types = "cTidcci")

  ## Check here to see if csv_df has any data in it
  if (nrow(csv_df) == 0) {
    stop("No data exists for this station query")
  }

  ## Rename columns to reflect tidyhydat naming
  colnames(csv_df) <- c("STATION_NUMBER","Date","Parameter","Value","Grade","Symbol","Approval")

  csv_df <- dplyr::left_join(
    csv_df,
    dplyr::select(tidyhydat.ws::param_id, -Name_Fr),
    by = c("Parameter")
  )
  csv_df <- dplyr::select(csv_df, STATION_NUMBER, Date, Name_En, Value, Unit, Grade, Symbol, Approval, Parameter, Code)

  ## What stations were missed?
  differ <- setdiff(unique(station_number), unique(csv_df$STATION_NUMBER))
  if (length(differ) != 0) {
    if (length(differ) <= 10) {
      message("The following station(s) were not retrieved: ", paste0(differ, sep = " "))
      message("Check station number for typos or if it is a valid station in the network")
    }
    else {
      message("More than 10 stations from the initial query were not returned. Ensure realtime and active status are correctly specified.")
    }
  } else {
    message("All station successfully retrieved")
  }

  ## Return it
  csv_df

  ## Need to output a warning to see if any stations weren't retrieved
  }


data-raw/process-param-id.R
# Copyright 2020 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.

library(readxl)

param_id <- read_excel('data-raw/parameterList_viaWebService.xlsx')

colnames(param_id) <- c("Parameter", "Code", "Unit", "Name_En", "Name_Fr",  "Description_En", "Description_Fr")

usethis::use_data(param_id, overwrite = TRUE)


R/webservice_functions.R
# Copyright 2017 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.



#' Request a token from the Environment and Climate Change Canada [EXPERIMENTAL]
#' @param username Supplied by ECCC. Defaults to NULL which will use WS_USRNM from .Renviron file
#' @param password Supplied by ECCC. Defaults to NULL which will use WS_PWD from .Renviron file
#' Request a token from the ECCC web service using the POST method. This token expires after 10 minutes.
#' You can only have 5 tokens out at once.
#'
#' @details The \code{username} and \code{password} should be treated carefully and should never be entered directly into an r script or console.
#' Rather these credentials should be stored in your \code{.Renviron} file. The .Renviron file can edited using \code{file.edit("~/.Renviron")}.
#' In that file, which is only stored locally and is only available to you, you can assign your \code{username} to WS_USRNM and \code{password}
#' to WS_PWD and then simply issue \code{token_ws}. See \code{?realtime_ws} for examples.
#'
#' @return The token as a string that should be supplied the \code{realtime_ws} function.
#'
#'
#' @family realtime functions
#' @export
#'


token_ws <- function(username = NULL, password = NULL) {

  ## Fail if the environmental variables aren't set
  if(!nzchar(Sys.getenv("WS_USRNM"))){
    stop("No environment variable for username assigned to WS_USRNM in .Renviron file", call. = TRUE)
  }
  if(!nzchar(Sys.getenv("WS_PWD"))){
    stop("No environment variable for password assigned to WS_PWD in .Renviron file", call. = TRUE)
  }


  if(is.null(username) | is.null(password)){
    username <- Sys.getenv("WS_USRNM")
    password <- Sys.getenv("WS_PWD")
  }

  login <- list(
    username = username,
    password = password
  )
  r <- httr::POST("https://wateroffice.ec.gc.ca/services/auth", body = login,
                  httr::user_agent(paste0("tidyhydat.ws v", utils::packageVersion("tidyhydat.ws"))))

  time_token <- Sys.time()

  ## A workaround that pauses for the connection
  #Sys.sleep(2)

  ## If the POST request was not a successful, print out why.
  ## Possibly could provide errors as per web service guidelines
  if (httr::status_code(r) == 422) {
    stop("422 Unprocessable Entity: Username and/or password are missing or are formatted incorrectly.")
  }

  if (httr::status_code(r) == 403) {
    stop("403 Forbidden: the web service is denying your request. Try any of the following options:
         -Ensure your credentials are set in your .Renviron file
         -Ensure you are not currently using all 5 tokens
         -Wait a few minutes and try again
         ")
  }

  ## Catch all error for anything not covered above.
  httr::stop_for_status(r)

  message(paste0("This token will expire at ", format(time_token + 10 * 60, "%H:%M:%S")))

  ## Extract token from POST
  token <- httr::content(r, "text", encoding = "UTF-8")

  attr(token, 'time') <- time_token

  return(token)
  }

#' Download realtime data from the ECCC web service
#'
#' Function to actually retrieve data from ECCC web service. Before using this function,
#' a token from \code{token_ws()} is needed. \code{realtime_ws} will let you know if the token has expired.
#' The maximum number of days that can be queried depends on other parameters being requested.
#' If one station is requested, 18 months of data can be requested. If you continually receiving
#' errors when invoking this function, reduce the number of observations (via station_number,
#' parameters or dates) being requested.
#'
#' @param station_number Water Survey of Canada station number.
#' @param parameters parameter ID. Can take multiple entries. Parameter is a numeric code. See \code{param_id}
#' for some options though undocumented parameters may be implemented. Defaults to Water level provisional, Secondary water level,
#' Tertiary water level, Discharge Provisional, Discharge, sensor, Water temperature, Secondary water temperature, Accumulated precipitation
#' @param start_date Accepts either YYYY-MM-DD or YYYY-MM-DD HH:MM:SS.
#' If only start date is supplied (i.e. YYYY-MM-DD) values are returned from the start of that day.
#' Defaults to 30 days before current date. Time is supplied in UTC.
#' @param end_date Accepts either YYYY-MM-DD or YYYY-MM-DD HH:MM:SS.
#' If only a date is supplied (i.e. YYYY-MM-DD) values are returned from the end of that day.
#' Defaults to current date. Time is supplied in UTC.
#' @param token generated by \code{token_ws()}
#'
#'
#' @format A tibble with 6 variables:
#' \describe{
#'   \item{STATION_NUMBER}{Unique 7 digit Water Survey of Canada station number}
#'   \item{Date}{Observation date and time. Formatted as a POSIXct class as UTC for consistency.}
#'   \item{Name_En}{Code name in English}
#'   \item{Value}{Value of the measurement.}
#'   \item{Unit}{Value units}
#'   \item{Grade}{future use}
#'   \item{Symbol}{future use}
#'   \item{Approval}{future use}
#'   \item{Parameter}{Numeric parameter code}
#'   \item{Code}{Letter parameter code}
#' }
#'
#' @examples
#' \dontrun{
#' token_out <- token_ws()
#'
#' ws_08 <- realtime_ws(station_number = c("08NL071","08NM174"),
#'                          parameters = c(47, 5),
#'                          token = token_out)
#'
#' fivedays <- realtime_ws(station_number = c("08NL071","08NM174"),
#'                          parameters = c(47, 5),
#'                          end_date = Sys.Date(), #today
#'                          start_date = Sys.Date() - 5, #five days ago
#'                          token = token_out)
#' }
#' @family realtime functions
#' @export


realtime_ws <- function(station_number, parameters = NULL,
                        start_date = Sys.Date() - 30, end_date = Sys.Date(), token) {
  #if (length(station_number) >= 300) {
  #  stop("Only 300 stations are supported for one request. If more stations are required,
  #       a separate request should be issued to include the excess stations. This second request can
  #       be issued on the same token if it isn't required.")
  #}

  if (is.null(parameters)) parameters <- c(46, 16, 52, 47, 8, 5, 41, 18)

  if (any(!parameters %in% param_id$Parameter)) {
    stop(paste0(paste0(parameters[!parameters %in% tidyhydat.ws::param_id$Parameter], collapse = ","),
                " are invalid parameters. Check param_id for a list of valid options."), call. = FALSE)
  }

  if (!is.numeric(parameters)) stop("parameters should be a number", call. = FALSE)


  ## Check to see if the token is expired
  time_token <- attr(token, 'time')
  if(format(time_token + 10 * 60) < Sys.time()){
    stop("Your token has expired. Retrieve a new one using token_ws()")
  }

  if(nchar(as.character(start_date)) == 10 | nchar(as.character(end_date)) == 10){
    start_date <- paste0(start_date, " 00:00:00")
    end_date <- paste0(end_date, " 23:59:59")
  }

  if(!grepl('[0-9]{4}-[0-1][0-9]-[0-3][0-9] [0-2][0-9]:[0-5][0-9]:[0-5][0-9]', start_date)) {
    stop("Invalid date format. start_date need to be in either YYYY-MM-DD or YYYY-MM-DD HH:MM:SS formats", call. = FALSE)
    }

  if(!grepl('[0-9]{4}-[0-1][0-9]-[0-3][0-9] [0-2][0-9]:[0-5][0-9]:[0-5][0-9]', end_date)) {
    stop("Invalid date format. start_date need to be in either YYYY-MM-DD or YYYY-MM-DD HH:MM:SS formats", call. = FALSE)
  }


  if(!is.null(start_date) & !is.null(end_date)){
    if (lubridate::ymd_hms(end_date) < lubridate::ymd_hms(start_date)) {
      stop("start_date is after end_date. Try swapping values.", call. = FALSE)
    }
  }



  ## Check date is in the right format
  if (is.na(as.Date(start_date, format = "%Y-%m-%d")) | is.na(as.Date(end_date, format = "%Y-%m-%d"))) {
    stop("Invalid date format. Dates need to be in YYYY-MM-DD format")
  }

  #if (as.Date(end_date) - as.Date(start_date) > 60) {
  #  stop("The time period of data being requested should not exceed 2 months.
  #       If more data is required, then a separate request should be issued to include a different time period.")
  #}
  ## English parameter names

  ## Is it a valid parameter name?

  ## Is it a valid Station name?


  ## Build link for GET
  baseurl <- "https://wateroffice.ec.gc.ca/services/real_time_data/csv/inline?"
  station_string <- paste0("stations[]=", station_number, collapse = "&")
  parameters_string <- paste0("parameters[]=", parameters, collapse = "&")
  date_string <- paste0("start_date=", substr(start_date, 1, 10), "%20", substr(start_date, 12, 19),
         "&end_date=", substr(end_date, 1, 10), "%20", substr(end_date, 12, 19))
  token_string <- paste0("token=", token[1])

  ## paste them all together
  url_for_GET <- paste0(
    baseurl,
    station_string, "&",
    parameters_string, "&",
    date_string, "&",
    token_string
  )

  ## Get data
  get_ws <- httr::GET(url_for_GET, httr::user_agent(paste0("tidyhydat.ws v", utils::packageVersion("tidyhydat.ws"))))

  ## Give webservice some time
  Sys.sleep(1)

  if (httr::status_code(get_ws) == 403) {
    stop("403 Forbidden: the web service is denying your request. Try any of the following options:
         -Ensure you are not currently using all 5 tokens
         -Wait a few minutes and try again
         -Copy the token_ws call and paste it directly into the console
         -Try using realtime_ws if you only need water quantity data
         ")
  }

  ## Check the GET status
  httr::stop_for_status(get_ws)

  if (httr::headers(get_ws)$`content-type` != "text/csv; charset=utf-8") {
    stop("GET response is not a csv file")
  }

  ## Turn it into a tibble and specify correct column classes
  csv_df <- httr::content(
    get_ws,
    type = "text/csv",
    encoding = "UTF-8",
    col_types = "cTidcci")

  ## Check here to see if csv_df has any data in it
  if (nrow(csv_df) == 0) {
    stop("No data exists for this station query")
  }

  ## Rename columns to reflect tidyhydat naming
  colnames(csv_df) <- c("STATION_NUMBER","Date","Parameter","Value","Grade","Symbol","Approval")

  csv_df <- dplyr::left_join(
    csv_df,
    dplyr::select(tidyhydat.ws::param_id, -.data$Name_Fr),
    by = c("Parameter")
  )
  csv_df <- dplyr::select(csv_df, .data$STATION_NUMBER, .data$Date, .data$Name_En, .data$Value, .data$Unit,
                          .data$Grade, .data$Symbol, .data$Approval, .data$Parameter, .data$Code)

  ## What stations were missed?
  differ <- setdiff(unique(station_number), unique(csv_df$STATION_NUMBER))
  if (length(differ) != 0) {
    if (length(differ) <= 10) {
      message("The following station(s) were not retrieved: ", paste0(differ, sep = " "))
      message("Check station number for typos or if it is a valid station in the network")
    }
    else {
      message("More than 10 stations from the initial query were not returned. Ensure realtime and active status are correctly specified.")
    }
  } else {
    message("All station successfully retrieved")
  }

  p_differ <- setdiff(unique(parameters), unique(csv_df$Parameter))
  if (length(p_differ) != 0) {
      message("The following valid parameter(s) were not retrieved for at least one station you requested: ", paste0(p_differ, sep = " "))
  } else {
    message("All parameters successfully retrieved")
  }


  ## Return it
  csv_df

  ## Need to output a warning to see if any stations weren't retrieved
  }
